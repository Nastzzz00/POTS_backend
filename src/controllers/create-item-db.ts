import { IDBModel } from '../commons/types';
import { createUCCreateItem } from '../core/';

const createCreateItemDB = (db: IDBModel<any>) => async i => {
  const createItem = createUCCreateItem();
  const newItem = createItem(i);
  return db.insert(newItem);
};

export { createCreateItemDB };
