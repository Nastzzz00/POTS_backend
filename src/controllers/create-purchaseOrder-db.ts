import { IDBModel } from '../commons/types';
import { createUCPurchaseOrder } from '../core/';

const createCreatePurchaseOrderDB = (db: IDBModel<any>) => async po => {
	const createPurchaseOrder = createUCPurchaseOrder();
	const newPurchaseOrder = createPurchaseOrder(po);
	return db.insert(newPurchaseOrder);
};

export { createCreatePurchaseOrderDB };
