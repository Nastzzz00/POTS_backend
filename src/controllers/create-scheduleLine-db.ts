import { IDBModel } from '../commons/types';
import { createUCCreateScheduleLine } from '../core/';

const createCreateScheduleLineDB = (db: IDBModel<any>) => async sl => {
	const createScheduleLine = createUCCreateScheduleLine();
	const newScheduleLine = createScheduleLine(sl);
	return db.insert(newScheduleLine);
};

export { createCreateScheduleLineDB };
