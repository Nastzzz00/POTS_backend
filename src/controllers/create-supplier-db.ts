import { IDBModel } from '../commons/types';
import { createUCCreateSupplier } from '../core/';

const createCreateSupplierDB = (db: IDBModel<any>) => async u => {
	const createSupplier = createUCCreateSupplier();
	const newSupplier = createSupplier(u);
	return db.insert(newSupplier);
};

export { createCreateSupplierDB };
