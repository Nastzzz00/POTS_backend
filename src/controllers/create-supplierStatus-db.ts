import { IDBModel } from '../commons/types';
import { createUCCreateSupplierStatus } from '../core/';

const createCreateSupplierStatusDB = (db: IDBModel<any>) => async u => {
	const createSupplierStatus = createUCCreateSupplierStatus();
	const newSupplierStatus = createSupplierStatus(u);
	return db.insert(newSupplierStatus);
};

export { createCreateSupplierStatusDB };
