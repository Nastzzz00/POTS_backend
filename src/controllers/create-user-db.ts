import { IDBModel } from '../commons/types';
import { createUCCreateUser } from '../core/';

const createCreateUserDB = (db: IDBModel<any>) => async u => {
	const createUser = createUCCreateUser();
	const newUser = createUser(u);
	return db.insert(newUser);
};

export { createCreateUserDB };
