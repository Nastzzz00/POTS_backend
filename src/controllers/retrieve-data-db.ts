import { IDBModel } from '../commons/types';

const getAllDataDB = (db: IDBModel<any>) => async () => {
	return db.getAll();
};
const getByIDDB = (db: IDBModel<any>) => async id => {
	return db.getById(id);
};

const getAllBySupplierStatusDB = (db: IDBModel<any>) => async id => {
	return db.getAllBySupplierStatus(id);
};

const getAllByItemDB = (db: IDBModel<any>) => async id => {
	return db.getAllByItem(id);
};

const getAllByScheduleLineDB = (db: IDBModel<any>) => async id => {
	return db.getAllByScheduleLine(id);
};

export { getAllDataDB, getByIDDB, getAllByItemDB, getAllBySupplierStatusDB, getAllByScheduleLineDB };
