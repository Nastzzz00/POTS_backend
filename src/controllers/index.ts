import {
  getAllDataDB,
  getByIDDB,
  getAllByItemDB,
  getAllBySupplierStatusDB,
  getAllByScheduleLineDB,
} from './retrieve-data-db';
import { DeleteRecordByIDDB } from './delete-data-db';

//User
import { createCreateUserDB } from './create-user-db';
import { updateUserByIDDB } from './update-user-db';

//Supplier
import { createCreateSupplierDB } from './create-supplier-db';
import { updateSupplierByIDDB } from './update-supplier-db';

//Supplier Status
import { createCreateSupplierStatusDB } from './create-supplierStatus-db';
import { updateSupplierStatusByIDDB } from './update-supplierStatus-db';
import { updateSupplierStatusItemByIDDB } from './update-supplierStatusItem-db';

//Item
import { createCreateItemDB } from './create-item-db';
import { updateItemByIDDB } from './update-item-db';

//Purchase Order
import { createCreatePurchaseOrderDB } from './create-purchaseOrder-db';
import { updatePurchaseOrderByIDDB } from './update-purchaseOrder-db';
import { updateAdminStatusByIDDB } from './update-adminStatus-db';

//Schedule Line
import { createCreateScheduleLineDB } from './create-scheduleLine-db';
import { updateScheduleLineByIDDB } from './update-scheduleLine-db';

export {
  getAllDataDB,
  getByIDDB,
  DeleteRecordByIDDB,
  getAllByItemDB,
  getAllBySupplierStatusDB,
  getAllByScheduleLineDB,
  //User exports
  createCreateUserDB,
  updateUserByIDDB,
  //Supplier exports
  createCreateSupplierDB,
  updateSupplierByIDDB,
  //SupplierStatus exports
  createCreateSupplierStatusDB,
  updateSupplierStatusByIDDB,
  //Item
  createCreateItemDB,
  updateItemByIDDB,
  updateSupplierStatusItemByIDDB,
  //Purchase Order
  createCreatePurchaseOrderDB,
  updatePurchaseOrderByIDDB,
  updateAdminStatusByIDDB,
  //Schedule Line
  createCreateScheduleLineDB,
  updateScheduleLineByIDDB,
};
