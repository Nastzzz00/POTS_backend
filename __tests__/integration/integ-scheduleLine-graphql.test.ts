import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import mockData from '../mocks/mock-data';
import * as controllers from '../../src/controllers';
import { constructTestServer } from '../__utils';

const {
  getAllDataDB,
  getByIDDB,
  DeleteRecordByIDDB,
  createCreateScheduleLineDB,
  updateScheduleLineByIDDB,
  getAllBySupplierStatusDB,
  createCreateSupplierStatusDB,
} = controllers;

const scheduleLinesMock = {
  insert: jest.fn(async input => {
    return { id: '1', ...input };
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.scheduleLines.filter(filterData);

    return res[0] || null;
  }),
  getAll: jest.fn(async () => {
    return mockData.scheduleLines;
  }),
  updateById: jest.fn(async input => {
    return { ...input };
  }),
  deleteById: async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.scheduleLines.filter(filterData);
    return res[0] || null;
  },
  getAllBySupplierStatus: jest.fn(async id => {
    const filterData = data => {
      if (id.includes(data._id)) {
        return data;
      }
    };
    const res = mockData.supplierStatus
      .filter(filterData)
      .map(data => ({ ...data, id: data._id }));
    return res;
  }),
  getAllByItem: async id => {},
  getAllByScheduleLine: async id => {},
  updateSupplierStatusItemById: async id => {},
  updateAdminStatusPurchaseOrderById: async id => {},
};

const supplierStatusMock = {
  insert: jest.fn(async input => {
    return { id: '1', ...input };
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.supplierStatus.filter(filterData);

    return res[0] || null;
  }),
  getAll: jest.fn(async () => {
    return mockData.supplierStatus;
  }),
  updateById: jest.fn(async input => {
    return { ...input };
  }),
  deleteById: async id => {},
  getAllBySupplierStatus: async id => {},
  getAllByItem: async id => {},
  getAllByScheduleLine: async id => {},
  updateSupplierStatusItemById: async id => {},
  updateAdminStatusPurchaseOrderById: async id => {},
};

const { server }: any = constructTestServer({
  context: {
    createScheduleLine: createCreateScheduleLineDB(scheduleLinesMock),
    updateScheduleLine: updateScheduleLineByIDDB(scheduleLinesMock),
    deleteScheduleLineById: DeleteRecordByIDDB(scheduleLinesMock),
    getScheduleLineById: getByIDDB(scheduleLinesMock),
    getAllScheduleLines: getAllDataDB(scheduleLinesMock),
    getAllSupplierStatusByScheduleLine: getAllBySupplierStatusDB(
      scheduleLinesMock,
    ),
    createSupplierStatus: createCreateSupplierStatusDB(supplierStatusMock),
  },
});

describe('Tests', () => {
  //Queries

  it('should fetch all schedule lines', async () => {
    const SCHEDULELINES_ALL = gql`
      query {
        allScheduleLines {
          id
          quantity
          uom
          unitPrice
          totalAmount
          deliveryDateAndTime
          deliveryStatus {
            id
            status
            dateCreated
            timeCreated
          }
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({ query: SCHEDULELINES_ALL });

    expect(res).toMatchSnapshot();
  });

  it('should fetch one schedule line', async () => {
    const SINGLE_SCHEDULELINE = gql`
      query ss($id: String!) {
        scheduleLine(id: $id) {
          id
          quantity
          uom
          unitPrice
          totalAmount
          deliveryDateAndTime
          deliveryStatus {
            id
            status
            dateCreated
            timeCreated
          }
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({
      query: SINGLE_SCHEDULELINE,
      variables: { id: '1' },
    });

    expect(res).toMatchSnapshot();
  });

  it('should error when no schedule line', async () => {
    const SINGLE_SCHEDULELINE = gql`
      query ss($id: String!) {
        scheduleLine(id: $id) {
          id
          quantity
          uom
          unitPrice
          totalAmount
          deliveryDateAndTime
          deliveryStatus {
            id
            status
            dateCreated
            timeCreated
          }
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({
      query: SINGLE_SCHEDULELINE,
      variables: { id: '' },
    });

    expect(res).toMatchSnapshot();
  });

  //Mutations

  it('create a schedule line', async () => {
    const CREATE_SCHEDULELINE = gql`
      mutation createSl($scheduleLine: ScheduleLineInput!) {
        createScheduleLine(scheduleLine: $scheduleLine) {
          id
          quantity
          uom
          deliveryDateAndTime
          unitPrice
          totalAmount
          deliveryStatus {
            id
            status
            dateCreated
            timeCreated
          }
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: CREATE_SCHEDULELINE,
      variables: {
        scheduleLine: {
          quantity: 10,
          uom: 'kilograms',
          deliveryDateAndTime: 'February 25,2020 4:30PM',
          unitPrice: 1000,
          totalAmount: 10000,
          deliveryStatus: [
            {
              status: 'Delivered',
            },
          ],
        },
      },
    });

    expect(res.errors).toBeUndefined();
    expect(scheduleLinesMock.insert.mock.calls.length).toBe(1);
    expect(res.data).toMatchObject({
      createScheduleLine: {
        id: '1',
        quantity: 10,
        uom: 'kilograms',
        deliveryDateAndTime: 'February 25,2020 4:30PM',
        unitPrice: 1000,
        totalAmount: 10000,
        deliveryStatus: [
          {
            id: '1',
            status: 'Dispatched',
            dateCreated: 'February 14, 2020',
            timeCreated: '4:30 PM',
          },
        ],
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete a schedule Line', async () => {
    const DELETE_SCHEDULELINE = gql`
      mutation sl($id: ID!) {
        deleteScheduleLine(id: $id) {
          id
        }
      }
    `;
    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: DELETE_SCHEDULELINE,
      variables: { id: 'U1' },
    });
    expect(res.errors).toBeUndefined();
    expect(res).toMatchSnapshot();
  });

  it('update a schedule line', async () => {
    const UPDATE_SCHEDULELINE = gql`
      mutation sl($scheduleLine: UpdateScheduleLineInput!) {
        updateScheduleLine(scheduleLine: $scheduleLine) {
          id
          deliveryStatus {
            id
            status
            dateCreated
            timeCreated
          }
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: UPDATE_SCHEDULELINE,
      variables: {
        scheduleLine: {
          id: '1',
          deliveryStatus: [
            {
              //   id: '1',
              status: 'Dispatched',
              //   dateCreated: 'February 14, 2020',
              //   timeCreated: '4:30 PM',
            },
          ],
        },
      },
    });

    expect(res).toMatchSnapshot();
  });
});
