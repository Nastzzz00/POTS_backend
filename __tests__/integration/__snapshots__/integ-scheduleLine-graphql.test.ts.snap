// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests create a schedule line 1`] = `
Object {
  "data": Object {
    "createScheduleLine": Object {
      "deliveryDateAndTime": "February 25,2020 4:30PM",
      "deliveryStatus": Array [
        Object {
          "dateCreated": "February 14, 2020",
          "id": "1",
          "status": "Dispatched",
          "timeCreated": "4:30 PM",
        },
      ],
      "id": "1",
      "quantity": 10,
      "totalAmount": 10000,
      "unitPrice": 1000,
      "uom": "kilograms",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests delete a schedule Line 1`] = `
Object {
  "data": Object {
    "deleteScheduleLine": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should error when no schedule line 1`] = `
Object {
  "data": Object {
    "scheduleLine": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should fetch all schedule lines 1`] = `
Object {
  "data": Object {
    "allScheduleLines": Array [
      null,
      null,
    ],
  },
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field ScheduleLine.id.],
    [GraphQLError: Cannot return null for non-nullable field ScheduleLine.id.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should fetch one schedule line 1`] = `
Object {
  "data": Object {
    "scheduleLine": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests update a schedule line 1`] = `
Object {
  "data": Object {
    "updateScheduleLine": null,
  },
  "errors": Array [
    [GraphQLError: Cannot read property 'quantity' of null],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
